motioneye/uploadservices.py

```diff
@@ -833,6 +834,87 @@ def _request_credentials(self, authorization_key):
         }
 
 
+class Nextcloud(UploadService):
+    NAME = 'nextcloud'
+
+    def __init__(self, camera_id):
+        self._server = None
+        self._port = None
+        self._username = None
+        self._password = None
+        self._location = None
+
+        UploadService.__init__(self, camera_id)
+
+    def _get_base_url(self):
+        scheme = 'http://'
+        if self._port == 443:
+            scheme = 'https://'
+        url = scheme + self._server + ':' + str(self._port) + '/remote.php/dav/files/' + self._username + '/'
+        return url
+
+    def _request(self, url, method, data=None):
+        self.debug('request: ' + method + ' ' + url)
+        request = urllib2.Request(url, data=data)
+        request.get_method = lambda: method
+        base64string = base64.b64encode('%s:%s' % (self._username, self._password))
+        request.add_header("Authorization", "Basic %s" % base64string)
+        if data is not None:
+            request.add_header('Content-Length', '%d' % len(data))
+        try:
+            utils.urlopen(request)
+        except urllib2.HTTPError as e:
+            if method == 'MKCOL' and e.code == 405:
+                self.debug('MKCOL failed with code 405, this is normal if the folder exists')
+            else:
+                raise e
+
+    def _make_dirs(self, path):
+        dir_url = self._get_base_url()
+        for folder in path.strip('/').split('/'):
+            dir_url = dir_url + folder + '/'
+            self._request(dir_url, 'MKCOL')
+
+    def test_access(self):
```

Comment:
This method returns a book or a str?

Reply:
Currently both. This is the same for other upload services, either `True` or a string on Exception. I haven't checked how it is used precisely and how returned error messages are handled, but probably it makes sense to consequently return a boolean and log or print the error message instead of returning it.

Reply:
So `if Nextcloud().test_access():` will ___always___ be True?

Reply:
`True` or a string. It is used (only) here: https://github.com/motioneye-project/motioneye/blob/c60b64a/motioneye/handlers/config.py#L578-L585

So if it result is `True`, a generic success message is logged, else the returned string is used in an error message, i.e. a string is evaluated as a failure.

Reply:
Marking this as "resolved". If we aim to change the methods to split error logging off and always have a boolean return here, we need to do it in a dedicated PR. This PR just follows the present logic/method, which is exactly how it should be.